src/api.rs

```diff
@@ -1347,45 +1369,49 @@ impl<T: Pixel> ContextInner<T> {
   }
 
   fn compute_lookahead_data(&mut self) {
-    // Process all new frames that we have received.
-    let frames_to_process = self
-      .frame_q
-      .iter()
-      .filter_map(|(&input_frameno, frame)| {
-        if input_frameno >= self.next_lookahead_frame {
-          frame.clone().map(|frame| (input_frameno, frame))
-        } else {
-          None
-        }
-      })
-      .collect::<Vec<_>>();
+    while self.lookahead_ready() {
+      // Process the next unprocessed frame
+      // Start by getting that frame and all frames after it in the queue
+      let lookahead_frames = self
+        .frame_q
+        .iter()
+        .filter_map(|(&input_frameno, frame)| {
+          if input_frameno >= self.next_lookahead_frame {
+            frame.clone()
+          } else {
+            None
+          }
+        })
+        .take(LOOKAHEAD_FRAMES as usize + 1)
+        .collect::<Vec<_>>();
```

Comment:
This is recomputing all lookahead frames to process every loop iteration, it seems that only the first one is needed.

Reply:
Ah, no, you're passing them to the keyframe detector... Maybe it's possible to compute them once, then, before the `while self.lookahead_ready()` loop?

Reply:
It might be possible. But I think this is not likely to incur a large performance penalty, and I think the code is cleaner this way. The `clone` initially throws red flags, but it's only cloning an `Arc` pointer, which only means incrementing the ref count.

Reply:
Well, I'm mainly worried about the `.collect::<Vec<_>>()`

Reply:
It would require a counter variable, but I think that's more performant than collecting each time, you're right. I'll push that.